File: cSharp/easypay/Signature.cs
namespace easypay.swisscom.com
{
    using System.Security.Cryptography;

    /// <summary>
    ///     Create <b>HmacSHA1</b> signature for data with key
    /// </summary>
    public class Signature
    {
        /// <summary>
        ///     create data-string for signature
        /// </summary>
        /// <param name="httpRequestMethod"> String </param>
        /// <param name="contentHash">       String </param>
        /// <param name="contentType">       String </param>
        /// <param name="date">              String </param>
        /// <param name="path">              String </param>
        /// <returns> String </returns>
        public static string CreateHashString(string httpRequestMethod, string contentHash, string contentType, string date, string path)
        {
            return httpRequestMethod + '\n' + contentHash + '\n' + contentType + '\n' + date + '\n' + path;
        }

        public static byte[] HashData(byte[] data)
        {
            using (var md5 = MD5.Create())
            {
                return md5.ComputeHash(data);
            }
        }

        /// <summary>
        ///     Create <b>HmacSHA1</b> byte[] signature for data and key
        /// </summary>
        /// <param name="key">  byte[] </param>
        /// <param name="data"> byte[] </param>
        /// <returns> signature byte[] </returns>
        public static byte[] Sign(byte[] key, byte[] data)
        {
            using (var hmac = new HMACSHA1(key))
            {
                // Compute the hash of hashString.
                return hmac.ComputeHash(data);
            }
        }
    }
}

File: javascript/easpay.lib/Signature.ts
import crypto = require('crypto');

export module easypay
{
    /**
    * Create <b>HmacSHA1</b> signature for data with key
    **/
    export class Signature
    {
        /**
        * create data-string for signature
        **/
        public static createHashString( httpRequestMethod : string,  contentHash : string,  contentType : string,  date: string,  path: string) : string
        {
            return httpRequestMethod + "\n" + contentHash + "\n" + contentType + "\n" + date + "\n" + path;
        }

		/**
		* Create <b>MD5</b> hash for data
		* @param {string} data
		* @returns {string} hash base64 encoded
		**/
        public static hashData(data: string) : string
        {
			var md5Sum = <any>crypto.createHash("md5");
			md5Sum.setEncoding("base64");
			md5Sum.write(data);
			md5Sum.end();
			var md5Hash = md5Sum.read();
			return md5Hash;
        }

        /**
        * Create <b>HmacSHA1</b> signature for data and key
		* @param {string} key
		* @param {string} data
		* @returns {string | Buffer} signature base64 encoded
        **/
        public static sign(key: string,  data: string) : string
        {
			var algorithm = "sha1";
            var hmac = crypto.createHmac(algorithm, key);
            hmac.setEncoding("base64");
            hmac.write(data);
            hmac.end();
            var shaHash = hmac.read();
			return <string>shaHash;
        }
    }
}
